### DISCUSSION TITLE
Strange behavior on cache update: cache data always null, still updates

### DISCUSSION THREAD

**Discussion by bartenra** [urql-graphql/urql#1729] [Q&A] [answered]
I use the following Graphcache update to update on a `visitorChat` subscription message:

```ts
Subscription: {
  visitorChat: (result, args, cache, info) => {
    console.log("received cache result", result);
    const { orgId, message } = result.visitorChat as ChatSubscriptionMessage;

    if (message) {
      cache.updateQuery({ query: UserChatsDocument }, data => {
        console.log("cache data", data);
        if (!data) return null;
        console.log("updating cache");
        let chat = data?.me?.chats.find(chat => chat.organization.id === orgId);
        chat?.messages.push({
          id: message.id,
          content: message.content,
          sender: message.sender,
          timestamp: message.timestamp,
          type: message.type
        });
        return data;
      });
    }
  }
}
```

The strange thing is, it always logs `cache data null`, but still the cache is updated. When I uncomment the line that pushes the new message into the chat (which it shouldn't reach when the cache `data` is null), it doesn't update anymore. So it almost seems like that line is run even when `data` is `null`! But... `updating cache` is never printed.

I'm really puzzled what's going on here. :see_no_evil:

**Comment by bartenra** [comment]
When I uncomment the `chat?.messages.push` line it *does* print `updating cache`. :smile:

**Comment by kitten** [comment, marked as answer]
that message you're pushing needs to contain the `__typename` for that particular object.

**Reply by bartenra** [reply]
OK I don't understand why, but that fixed it. :)

**Reply by JoviDeCroock** [reply]
well the _why_ is because we need a way to normalize it, we can't assume the type out of nowhere, the GraphQL api will also return you the `__typename` how normalization happens is explained in the documentation https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/
